structuredhub.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package pubsub
+
+import (
+	"encoding/json"
+	"reflect"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+)
+
+type structuredHub struct {
+	simplehub
+
+	marshaller  Marshaller
+	annotations map[string]interface{}
+	postProcess func(map[string]interface{}) (map[string]interface{}, error)
+}
+
+// Marshaller defines the Marshal and Unmarshal methods used to serialize and
+// deserialize the structures used in Publish and Subscription handlers of the
+// structured hub.
+type Marshaller interface {
+	Marshal(interface{}) ([]byte, error)
+	Unmarshal([]byte, interface{}) error
+}
+
+// StructuredHubConfig is the argument struct for NewStructuredHub.
+type StructuredHubConfig struct {
+	// Marshaller defines how the structured hub will convert from structures to
+	// a map[string]interface{} and back. If this is not specified, the
+	// `JSONMarshaller` is used.
+	Marshaller Marshaller
+
+	// Annotations are added to each message that is published if and only if
+	// the values are not already set.
+	Annotations map[string]interface{}
+
+	// PostProcess allows the caller to modify the resulting
```

Comment:
Can you expand on why this exists? (I think I know but I think the docstring should explain)